pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -81,27 +85,75 @@ func PodChanged(oldPod, newPod *v1.Pod, endpointChanged EndpointsMatch) (bool, b
 
 // GetPodServiceMemberships returns a set of Service keys for Services that have
 // a selector matching the given pod.
-func GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
+func GetPodServiceMemberships(serviceLister v1listers.ServiceLister, selectorCache *sync.Map, pod *v1.Pod) (sets.String, error) {
 	set := sets.String{}
-	services, err := serviceLister.GetPodServices(pod)
+	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
 	if err != nil {
 		// don't log this error because this function makes pointless
 		// errors when no services match
 		return set, nil
 	}
-	for i := range services {
-		key, err := controller.KeyFunc(services[i])
-		if err != nil {
-			return nil, err
+
+	var selector labels.Selector
+	for _, service := range services {
+		if v, ok := selectorCache.Load(fmt.Sprintf("%s_%s", service.Namespace, service.Name)); ok {
+			selector = v.(labels.Selector)
+		} else {
+			if service.Spec.Selector == nil {
+				// if the service has a nil selector this means selectors match nothing, not everything.
+				continue
+			}
+			selector = labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+			selectorCache.Store(fmt.Sprintf("%s_%s", service.Namespace, service.Name), selector)
+		}
+		if selector.Matches(labels.Set(pod.Labels)) {
+			key, err := controller.KeyFunc(service)
+			if err != nil {
+				return nil, err
+			}
+			set.Insert(key)
 		}
-		set.Insert(key)
 	}
 	return set, nil
 }
 
+// SyncSelectorCache will confirm consistency between local selectorCache and service's selector.
+func SyncSelectorCache(service *v1.Service, selectorCache *sync.Map) {
+	v, ok := selectorCache.Load(fmt.Sprintf("%s_%s", service.Namespace, service.Name))
+	if !ok {
+		return
+	}
+
+	oldSelector := v.(labels.Selector).String()
+	curSelector := selectorToString(service.Spec.Selector)
```

Comment:
There's a very helpful `selector.Matches()` function that could remove the need for this. Example here: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/validation/validation.go#L70

Reply:
All addressed, thanks for your review and excellent suggestions.

just one dissent: selector compare should use Equal rather than Matches. Matches can't catch change from [k1:v1, k2:v2] to [k1:v1]

Reply:
You're completely right, thanks for catching that!